### DISCUSSION TITLE
Force Charge Time - automatically set based on Agile target windows

### DISCUSSION THREAD

**Discussion by nicwats** [wills106/homeassistant-solax-modbus#789] [General]
Hi All, 
Apologies if it's covered elsewhere but I haven't been able to find the answer. I have everything set up and running fine with Home Assistant & Solax Modbus and have been using bottlecapdave's Octopus Energy addon to identify the optimal overnight charging window for the batteries. This is done by creating an entity which switches 'On' at the start of the charging window and remains on until the end of the window. 
I'm struggling to work out how to use the start of that 'On' state as the trigger to enter that particular time as the 'Solax Charger Start Time 1'. It's probably something very simple but when I look at it as an automation it will only let me amend the option of the 'Charger Start Time 1' via selecting a drop down time which is no good to me. 
I'm sure others have this working properly so hopefully it's a quick answer for me! 
Thanks!
<img width="1042" alt="Screenshot 2024-03-04 at 15 01 17" src="https://github.com/wills106/homeassistant-solax-modbus/assets/108125210/65cb264a-e68a-4f22-b223-44b0bb51e7b2">

**Comment by nicwats** [comment]
Might not be the most elegant solution but I think I may have found a way to make it work in case it helps others. 
I have two entities created using the bottlecapdave Octopus Energy addon which identify the optimal 3 hour window. One overnight (charge to 95% at cheap rate) and the other on days (I want the day time one to top up the battery to 50% prior to the evening peak to make sure I don't draw from the grid at all). 
I then set the two 'Force Charge Windows', one as a blanket duration covering the whole of the possible night window and the other covering the day window with their respective target rates of charge. 
I then set up a Home Assistant automation to switch between Self Use and Force Time Use depending on the on/off state of the charging window entity. Seems to work just fine but happy to take any pointers if I've missed anything or it could be further refined.

**Comment by neildhunter** [comment]
I've approached this from a slightly different angle! Rather than trying to update the charging schedule at the point the cheapest window becomes _active_ I use the point that the cheapest window is _updated_. So when the next day's charges are published the target binary sensor _next_time_ attribute changes - and at that point I update the Inverter's settings with the new start and end time.
That way I can leave the Inverter in self-use all the time and just shift the start and end times accordingly.

**Reply by nicwats** [reply]
Any chance you'd mind sharing your coding associated with it? In principle mine should work, and it has worked a few times now, but I also keep having an issue where the automation triggers it into 'Force Time' mode but then around 20 seconds later it reverts to 'Self Use' without any input seeming to have caused it. I'm wondering if it's some other app or something else disrupting it but am a little stumped at the moment. It might be that a different approach fixes things.

**Reply by neildhunter** [reply]
It could also be that if the Inverter is in "Idle" mode then it won't accept changes...
My stuff is here - help yourself!
https://github.com/neildhunter/HomeAssistant-Public/tree/main/SolaX%20Charge%20Time
